### PR TITLE
[DDW-262] Refactor Environment Setup

### PR DESCRIPTION
**Pull request by MarcusHurney**
This is a major refactoring of the render process, away from electrons `nodeIntegration` and `remote` apis, to a more secure `preload` script that provides a minimal compatibility layer to the frontend to work with some required node packages over a simple api surface. This is the recommended electron setup with focus on security.

The second large change is a cleanup of the environment setup to be more structured and easier to read. Of course also the environment does not mix node / browser context anymore and is just constructed once on the main process and passed to the renderer via the preload script.

To make this possible, many changes had to be made on the way:

### Improvements

- New web preferences for the main window that establish the new security layer:
(*note:* nodeIntegration and remote modules are only available during testing since that's a requirement for Spectron)
```js
 webPreferences: {
      nodeIntegration: isTest,
      webviewTag: false,
      enableRemoteModule: isTest,
      preload: path.join(__dirname, './preload.js')
 }
```
- Modified webpack compilation so that `target` depends on the environment (`isTestEnv ? 'electron-renderer' : 'web'`) because Spectron needs node integration.
- Environment setup was cleaned up (E.g: getters like `isMainnet()` are now simple vars `isMainnet` instead)
- Preload script exposes the environment vars via `global.environment`
- Any store has a new property `environment` which is a reference to `global.environment`
- The environment vars are not referenced directly in the components anymore but always passed down from containers using the new store property
- Compilation time reduced to ~14 seconds because of the refactoring of paper wallets logic to main process.
- Test suite is also a bit faster because of the smaller code size of the renderer process (it was cut in half without the paper wallet stuff)

### Changes

- Refactored the logic to generate a paper wallet to main process since that requires node.js
- Refactored send bug report request logic to main process since that requires node.js
- Modified the main webpack config to handle font files correctly

### Fixes

- Fixed the compilation warnings that were caused by missing, optional dependencies of `electron-connect`.

## TODO (ALL DONE)

**Very important:**
- [x] fix broken show save dialogs (replacing `remote.dialog.showSaveDialog` which requires node integration) (like seen in `/BugReportDialogContainer.js` and a few other places)
- [x] Refactor `reportRequest.js` to the main process (or at least the parts that require node integration features)
- [x] Refactor `paperWalletPdfGenerator.js` to main process since it requires node integration

**check:**
- [x] Fix broken `electron-log` integration via preload script (doesn't seem to work correctly atm)
- [x] Check if electron-store works correctly when exposed via preload script
- [x] Find out why requests to `/api/internal/next-update` are returning 404 errors
- [x] Find a more secure way to handle CORS issues with TLS certs -> currently they are just "accepted" without any check by basically disabling that Chrome security feature!

### IMPORTANT:
A few major refactorings have been done which need to be properly reviewed!

- [x] Generating paper wallets (was moved from renderer to main process)
- [x] Sending bug reports (was moved from renderer to main process)
- [ ] Preload script is compiled to `dist/main` output folder

### Basics

- [ ] PR is updated to the most recent version of target branch (and there are no conflicts)
- [ ] PR has good description that summarizes all changes and shows some screenshots or animated GIFs of important UI changes
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance tests are passing (`yarn run test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn run dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn run package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn run flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn run lint`)
- [ ] Text changes are proofread and approved (Jane Wild)
- [ ] There are no missing translations (running `yarn run manage:translations` produces no changes)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn run storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly documented and commented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-rendering
- [ ] Any code that only works in Electron is neatly separated from components

### Testing
- [ ] New feature / change is covered by acceptance tests
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature / change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review:
- [ ] Merge PR
- [ ] Delete source branch
- [ ] Move ticket to `done` on the Youtrack board

### CODE REVIEW

File: source/main/utils/preload.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    platform: _process.platform
+  };
+  global.Buffer = Buffer;
+  global.require = require;
+  // ESLint will warn about any use of eval(), even this one
+  // eslint-disable-next-line
+  global.eval = () => {
+    throw new Error('This app does not support window.eval().');
+  };
+});
```
Comment (DominikGuzei) [inline source/main/utils/preload.js]: I really like the preload script idea … why not copy the whole environment here and pick it up in `environment.js` if it exists? This way we could remove the IPC channel and weird sprinkle of environment logic all over the stores (which binds the stores to IPC even more)

**Comment by DominikGuzei** [comment]
@MarcusHurney please continue to work on this PR! I added a TODO list to the PR description, there are still many small / medium sized issues to be fixed to make Daedalus work without node integration. BUT we are closer to that goal than ever before! 🎉

**Comment by MarcusHurney** [comment]
@DominikGuzei 
Ok, will do! I'll start back on it Monday. I updated to Mojave because the latest version of nix supports it now. After updating, I needed to rebuild `cardano-sl` and it took 6+ hours 😮

**Comment by DominikGuzei** [comment]
@nikolaglumac @MarcusHurney this is now ready for final review 🙏

**Comment by DominikGuzei** [comment]
@nikolaglumac can you try it with another node version **and** within nix-shell?

**Pull request merged**